[SPARK-44647][SQL] Support SPJ where join keys are less than cluster keys

szehon-ho opened the pull request:
### What changes were proposed in this pull request?
- Add new conf spark.sql.sources.v2.bucketing.allowJoinKeysSubsetOfPartitionKeys.enabled
- Change key compatibility checks in EnsureRequirements.  Remove checks where all partition keys must be in join keys to allow isKeyCompatible = true in this case (if this flag is enabled)
- Change BatchScanExec/DataSourceV2Relation to group splits by join keys if they differ from partition keys (previously grouped only by partition values).  Do same for all auxiliary data structure, like commonPartValues.
- Implement partiallyClustered skew-handling.
  - Group only the replicate side (now by join key as well), replicate by the total size of other-side partitions that share the join key.
  - add an additional sort for partitions based on join key, as when we group the replicate side, partition ordering becomes out of order from the non-replicate side.

### Why are the changes needed?
- Support Storage Partition Join in cases where the join condition does not contain all the partition keys, but just some of them

### Does this PR introduce _any_ user-facing change? 
No

### How was this patch tested? 
-Added tests in KeyGroupedPartitioningSuite
-Found two existing problems, will address in separate PR:
- Because of https://github.com/apache/spark/pull/37886   we have to select all join keys to trigger SPJ in this case, otherwise DSV2 scan does not report KeyGroupedPartitioning and SPJ does not get triggered.  Need to see how to relax this.
- https://issues.apache.org/jira/browse/SPARK-44641 was found when testing this change.  This pr refactors some of those code to add group-by-join-key, but doesnt change the underlying logic, so issue continues to exist.  Hopefully this will also get fixed in another way.

sunchao commented on sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/BatchScanExec.scala:
```diff
@@ -144,8 +159,25 @@ case class BatchScanExec(
-              groupSplits = true).get
+            // In the case where we replicate partitions, we have grouped
+            // the partitions by the join key if they differ
+            val groupByExpressions =
```
Can we override `KeyGroupedPartitioning` method in this class, and wrap the logic of handling join keys in the method? We can return a new `KeyGroupedPartitioning` instance whose `expressions`, `partitionValues` are "projected" on the join keys.

szehon-ho replied:
Done, changed outputPartitioning to return KeyGroupedPartitoning to reflect that.

sunchao commented on sql/core/src/main/scala/org/apache/spark/sql/execution/exchange/EnsureRequirements.scala:
```diff
@@ -523,23 +546,25 @@ case class EnsureRequirements(
-      node.mapChildren(child => populatePartitionValues(
-        child, values, applyPartialClustering, replicatePartitions))
+      node.mapChildren(child => populateStoragePartitionJoinParams(
+        child, values, partitionGroupByPositions, applyPartialClustering, replicatePartitions))
```
Instead of populating `partitionGroupByPositions`, can we populate `StoragePartitionJoinParams.keyGroupedPartitioning` instead? which can be the subset of expressions that participate in the join.

szehon-ho replied:
I will need some more guidance on this one.

szehon-ho commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/physical/partitioning.scala:
```diff
@@ -711,36 +718,82 @@ case class KeyGroupedShuffleSpec(
+
+object KeyGroupedShuffleSpec {
+
+  def isExpressionCompatible(left: Expression, right: Expression): Boolean =
```
This is just grouping the new static into companion object, so the diff looks a bit bigger, let me know if I should revert

szehon-ho commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/physical/partitioning.scala:
```diff
@@ -701,41 +705,78 @@ case class KeyGroupedShuffleSpec(
+  }
+
+  // Whether the partition keys (i.e., partition expressions) that also are in the set of
+  // cluster keys are compatible between this and the 'other' spec.
```
I think its moved from existing javadoc, but removed it.

szehon-ho commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/physical/partitioning.scala:
```diff
@@ -701,41 +705,78 @@ case class KeyGroupedShuffleSpec(
+    partitioning.partitionValues.zip(otherPartitioning.partitionValues)
+      .forall {
+        case (left, right) =>
+          val leftTypes = partitioning.expressions.map(_.dataType)
```
Got rid of this actually.

szehon-ho commented on sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:
```diff
@@ -1500,6 +1500,18 @@ object SQLConf {
+      )
+      .version("3.5.0")
+      .booleanConf
+      .createWithDefault(true)
```
Done

szehon-ho commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/physical/partitioning.scala:
```diff
@@ -344,7 +344,11 @@ case class KeyGroupedPartitioning(
+            // Support only when all cluster key have an associated partition expression key
+            requiredClustering.exists(x => attributes.exists(_.semanticEquals(x))) &&
+              // and if all partition expression contain only a single partition key.
+               expressions.forall(_.collectLeaves().size == 1)
```
This was to fix a test, I couldn't find it back to be honest.  There was a test somewhere that was trying this case (which isnt actually supported in the code currently), and I think asserting the right exception, which I think would break if SPJ is activated.  I could revert this and see again to find the test, if you want.

dongjoon-hyun commented:
Could you re-trigger the failed pipeline or rebase this PR to the `master` branch once more, @szehon-ho ?

szehon-ho commented:
Hi @dongjoon-hyun , I think @sunchao had another idea he is thinking about, was going to wait a bit for that to update the pr

dongjoon-hyun commented:
Oh, got it!

szehon-ho committed:
[SPARK-44647][SQL] Support SPJ where join keys are less than cluster keys

 ### What changes were proposed in this pull request?
- Add new conf spark.sql.sources.v2.bucketing.allowJoinKeysSubsetOfPartitionKeys.enabled
- Change key compatibility checks in EnsureRequirements.  Remove checks where all partition keys must be in join keys to allow isKeyCompatible = true in this case (if this flag is enabled)
- "Project" partitions by join keys in KeyGroupedPartitioning/KeyGroupedShuffleSpec
- Add join key grouping to the partition grouping in BatchScanExec

   ### Why are the changes needed?
- Support Storage Partition Join in cases where the join condition does not contain all the partition keys, but just some of them

    ### Does this PR introduce _any_ user-facing change?
No

    ### How was this patch tested?
-Added tests in KeyGroupedPartitioningSuite
-Because of https://github.com/apache/spark/pull/37886   we have to select all join keys to trigger SPJ in this case, otherwise DSV2 scan does not report KeyGroupedPartitioning and SPJ does not get triggered.  Need to see how to relax this in separate PR.

sunchao commented on sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:
```diff
@@ -1530,6 +1530,18 @@ object SQLConf {
+      .doc("Whether to allow storage-partition join in the case where join keys are" +
+        "a subset of the partition keys of the source tables.  At planning time, " +
+        "Spark will group the partitions by only those keys that are in the join keys." +
+        "This is currently enabled only if spark.sql.sources.v2.bucketing.pushPartValues.enabled " +
```
nit: this is out-dated.

sunchao commented on sql/core/src/test/scala/org/apache/spark/sql/connector/KeyGroupedPartitioningSuite.scala:
```diff
@@ -1276,4 +1279,266 @@ class KeyGroupedPartitioningSuite extends DistributionAndOrderingSuiteBase {
+                "ON t1.id = t2.id ORDER BY t1.id, t1data, t2data")
+
+            // Currently SPJ for case where join key not same as partition key
+            // only supported when push-part-values enabled
```
nit: the comment is out-dated

sunchao reviewed (commented):
Thanks @szehon-ho ! Looks great with a few comments.

szehon-ho committed:
Review comments

szehon-ho committed:
Review comments

szehon-ho commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/physical/partitioning.scala:
```diff
@@ -344,7 +344,14 @@ case class KeyGroupedPartitioning(
+            if (SQLConf.get.getConf(
+              SQLConf.V2_BUCKETING_ALLOW_JOIN_KEYS_SUBSET_OF_PARTITION_KEYS)) {
+              requiredClustering.exists(x => attributes.exists(_.semanticEquals(x))) &&
+                expressions.forall(_.collectLeaves().size == 1)
```
Added some comment, please check if it makes sense

szehon-ho commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/physical/partitioning.scala:
```diff
@@ -674,7 +711,8 @@ case class HashShuffleSpec(
     partitioning: KeyGroupedPartitioning,
-    distribution: ClusteredDistribution) extends ShuffleSpec {
+    distribution: ClusteredDistribution,
+    joinKeyPositions: Option[Seq[Int]] = None) extends ShuffleSpec {
```
Added comments, please check and suggest if it can be improved.

szehon-ho commented:
@sunchao thanks! addressed review comments

sunchao commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/physical/partitioning.scala:
```diff
@@ -672,9 +708,17 @@ case class HashShuffleSpec(
 
+/**
+ * [[ShuffleSpec]] created by [[KeyGroupedPartitioning]].
+ * @param partitioning key grouped partitioning
```
nit: leave an empty line above the first `@param`

sunchao commented on sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:
```diff
@@ -1530,6 +1530,18 @@ object SQLConf {
+      .doc("Whether to allow storage-partition join in the case where join keys are" +
+        "a subset of the partition keys of the source tables.  At planning time, " +
+        "Spark will group the partitions by only those keys that are in the join keys." +
+        "This is currently enabled only if spark.sql.requireAllClusterKeysForDistribution " +
```
nit: replace with `spark.sql.requireAllClusterKeysForDistribution` `${REQUIRE_ALL_CLUSTER_KEYS_FOR_DISTRIBUTION.key}` (and add `s` to the beginning of this line)

sunchao reviewed (approved):
LGTM

dongjoon-hyun commented on sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:
```diff
@@ -1530,6 +1530,18 @@ object SQLConf {
+        "Spark will group the partitions by only those keys that are in the join keys." +
+        "This is currently enabled only if spark.sql.requireAllClusterKeysForDistribution " +
+        "is false."
+      )
```
nit. Let's put `)` at the end of the previous line.
```scala
"is false.")
```

dongjoon-hyun commented on sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:
```diff
@@ -1530,6 +1530,18 @@ object SQLConf {
+  val V2_BUCKETING_ALLOW_JOIN_KEYS_SUBSET_OF_PARTITION_KEYS =
+    buildConf("spark.sql.sources.v2.bucketing.allowJoinKeysSubsetOfPartitionKeys.enabled")
+      .doc("Whether to allow storage-partition join in the case where join keys are" +
+        "a subset of the partition keys of the source tables.  At planning time, " +
```
nit. `.  At` -> `. At`.

szehon-ho committed:
Fix typo

szehon-ho committed:
Fix sqlconf test

dongjoon-hyun reviewed (approved):
+1, LGTM.

dongjoon-hyun commented:
Merged to mater for Apache Spark 4.0.0. Thank you so much, @szehon-ho and @sunchao !

sunchao commented:
Thanks @szehon-ho @dongjoon-hyun !

irsath commented:
Hi @dongjoon-hyun @sunchao,
Do you see any blocker to backport this to spark 3.5 ?
I think it would be useful for many use case (including mine) that partition by date for GDPR purpose but still need SPJ on the other partitioning column.

dongjoon-hyun commented:
Apache Spark has a back-porting policy which allows only bug fixes, @irsath . Given that this PR is an improvement, we are unable to touch the release branches like `branch-3.5` for this improvement.

irsath commented:
Right, sorry for the typo but I meant: what if we make a 3.6 with this PR ?

I never contributed to OSS spark but if your ok with the idea I can try to do a PR in that regard.
